### ISSUE TITLE
To log or track output during training of NER model?

### ISSUE THREAD

**Issue by prashant334** [explosion/spaCy#1866]
Python version: 2.7.6
Platform: Linux-3.16.0-77-generic-x86_64-with-Ubuntu-14.04-trusty
spaCy version: 2.0.0a17
Models: en, en_core_web_sm, xx_ent_wiki_sm

@honnibal 

As you have mentioned in https://github.com/explosion/spaCy/issues/1864 that "**The short answer to your question is that the NER currently uses 4 CNN layers with residual connections and the maxout activation function to perform the "encode" step.** "

I want to track and  generate log during training of ner model , so that I can learn how entire processi is happened from sentence level to prediction level , any solution is there? 
please help......

**Comment by honnibal** [comment]
I would suggest checking out a copy of the code, and compiling from source. It may also be useful to check out a copy of `thinc`, the neural network library, and running that copy. Alternatively you can edit the copy of `thinc` installed in your site packages in your spaCy installation. Most of the relevant code is in pure Python, so it's easy to modify and run.

During training of the NER model, calls are made to the `Parser.update()` method here: https://github.com/explosion/spaCy/blob/v2.0.5/spacy/syntax/nn_parser.pyx#L545 . This update method is called with a batch of documents and their gold-standard entities. The call is usually made by `Language.update()`, but you can also call the method directly.

The `Parser.update()` method assigns vectors to each token using the convolutional neural network. The CNN is constructed in `Parser.Model()`. The architecture is defined here: https://github.com/explosion/spaCy/blob/v2.0.5/spacy/_ml.py#L257 . Once the token vectors are predicted by the CNN, the positional features are extracted, by pre-computing the output of the first hidden layer of the network. This is done in the `precompute_hiddens` class defined within `nn_parser.pyx`. The relevant neural network layer is defined here: https://github.com/explosion/spaCy/blob/v2.0.5/spacy/_ml.py#L138 . Finally the parser steps through the states of the sentence, predicting the next action at each point. This part of the training is somewhat complicated, as we must calculate the cost of each action in terms of the number of new errors the action introduces. This allows the gradient of the loss for each action to be computed.

Once the gradients are computed for each action, the loss is backpropagated to produce a gradient for each state feature. This is backpropagated through the first hidden layer to produce the gradient of the loss for each token vector. Finally, the loss on each token vector is backpropagated through the token CNN and into the embedding tables. After the loss is entirely backpropagated, the optimizer is called with each `(weight, gradient)` pair for the batch.
